### PR TITLE
Add a Math Library for Combining Pyth Prices on Solidity

### PR DESCRIPTION
**Pull request by nirala-mehul**
This PR adds the Price Library, a Solidity library comibing prices using `Price` data types. The library includes functions for basic arithmetic operations, normalization, scaling, and fraction calculations on Price structs.

## Changes Made
- **Price Library:** Added a new Solidity library (PriceLibrary.sol) that provides essential operations for manipulating Pyth prices.
- **Forge Test Project:** Introduced a test project (forge-test) to validate the functionality of the Price Library across various scenarios.
- **Test Cases:** Included comprehensive test cases within the forge-test project to ensure robustness and accuracy of the implemented functions.

## Summary of Public Functions Implemented in Price Library

- **getPriceInQuote**: Calculates the price of a base asset in terms of a quote asset using division and scaling.
- **getCollateralValuationPrice**: Computes an adjusted price based on deposits, applying rate discounts linearly interpolated between initial and final rates.
- **getBorrowValuationPrice**: Computes an adjusted price based on borrows, applying rate premiums linearly interpolated between initial and final premiums.
- **affineCombination**: Performs an affine combination of two prices at a query point, useful for interpolation.
- **normalize**: Normalizes a Price struct by adjusting its price and confidence based on its exponent.
- **scaleToExponent**: Scales a Price struct to a target exponent, adjusting its price and confidence accordingly.
- **div**: Performs division between two Price structs, adjusting for precision and handling overflow conditions.
- **add**: Adds two Price structs together, ensuring their exponents match.
- **cmul**: Performs constant multiplication of a Price struct by a scalar constant.
- **mul**: Multiplies two Price structs together, adjusting their exponents and handling overflow.

Testing
Added a forge project and tests for PriceLibrary.

### CODE REVIEW

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        PythStructs.Price memory discountInterpolated = affineCombination(
+            0,
+            initialPercentage,
+            int64(depositsEndpoint),
+            finalPercentage,
+            int64(deposits),
+            -9
+        );
+        // If interpolated price is zero, return zero price
+        if (discountInterpolated.price == 0) {
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:87]: would not do this. `confidence` could still be nonzero

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+            return PythStructs.Price(0, 0, 0, 0);
+        }
+
+        // Apply discounted price to self price
+        PythStructs.Price memory priceDiscounted = mul(self, discountInterpolated);
+        // If resulting price is zero, return zero price
+        if (priceDiscounted.price == 0) {
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:94]: would not do this. `conf` could still be nonzero

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        PythStructs.Price memory premiumInterpolated = affineCombination(
+            0,
+            initialPercentage,
+            int64(borrowsEndpoint),
+            finalPercentage,
+            int64(borrows),
+            -9
+        );
+        // If interpolated price is zero, return zero price
+        if (premiumInterpolated.price == 0) {
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:154]: same point as above, would not do this

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+        PythStructs.Price memory self,
+        uint64 borrows,
+        uint64 borrowsEndpoint,
+        uint64 ratePremiumInitial,
+        uint64 ratePremiumFinal,
+        int32 premiumExponent
+    ) public pure returns (PythStructs.Price memory) {
+        // If initial premium rate is greater than final rate, return zero price
+        if (ratePremiumInitial > ratePremiumFinal) {
+            return PythStructs.Price(0, 0, 0, 0);
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:127]: return error here

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+        int64 x1,
+        PythStructs.Price memory y1,
+        int64 x2,
+        PythStructs.Price memory y2,
+        int64 xQuery,
+        int32 preAddExpo
+    ) public pure returns (PythStructs.Price memory) {
+        // If x2 <= x1, return zero price
+        if (x2 <= x1) {
+            return PythStructs.Price(0, 0, 0, 0);
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:194]: throw error

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+
+        // Calculate deltas and fractions
+        int64 deltaQ1 = xQuery - x1;
+        int64 delta2Q = x2 - xQuery;
+        int64 delta21 = x2 - x1;
+
+        PythStructs.Price memory fracQ1 = fraction(deltaQ1, delta21);
+        // If fraction price is zero, return zero price
+        if (fracQ1.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:205]: i don't understand why you're returning 0 price here

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+        // If fraction price is zero, return zero price
+        if (frac2Q.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
+        }
+
+        // Perform multiplication and scaling to target exponent
+        PythStructs.Price memory left = mul(y2, fracQ1);
+        // If resulting price is zero, return zero price
+        if (left.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:217]: return here doesn't make sense

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+
+        PythStructs.Price memory fracQ1 = fraction(deltaQ1, delta21);
+        // If fraction price is zero, return zero price
+        if (fracQ1.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
+        }
+        PythStructs.Price memory frac2Q = fraction(delta2Q, delta21);
+        // If fraction price is zero, return zero price
+        if (frac2Q.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:210]: ditto

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+        // Perform multiplication and scaling to target exponent
+        PythStructs.Price memory left = mul(y2, fracQ1);
+        // If resulting price is zero, return zero price
+        if (left.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
+        }
+        PythStructs.Price memory right = mul(y1, frac2Q);
+        // If resulting price is zero, return zero price
+        if (right.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:222]: ditto

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+        }
+
+        left = scaleToExponent(left, preAddExpo);
+        // If scaled price is zero, return zero price
+        if (left.price == 0) {
+            return PythStructs.Price(0, 0, 0, 0);
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:228]: none of the early returns makes sense

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+        (uint64 basePrice, int64 baseSign) = toUnsigned(base.price);
+        (uint64 otherPrice, int64 otherSign) = toUnsigned(normalizedOther.price);
+
+        // Calculate midprice and exponent for division
+        uint64 midprice = (basePrice * PD_SCALE) / otherPrice;
+        int32 midpriceExpo = base.expo - normalizedOther.expo + PD_EXPO;
+
+        // Calculate confidence and handle overflow conditions
+        uint64 otherConfidencePct = (normalizedOther.conf * PD_SCALE) / otherPrice;
+        uint128 conf = ((uint128(base.conf) * uint128(PD_SCALE)) /
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:269]: i think your parentheses here are off here

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 12]
```diff
@@ -0,0 +1,481 @@
...
+        if (conf < type(uint64).max) {
+            return PythStructs.Price({
+                price: int64(midprice),
+                conf: uint64(conf),
+                expo: midpriceExpo,
+                publishTime: base.publishTime
+            });
+        } else {
+            // Otherwise, return zero Price struct
+            return PythStructs.Price({price: 0, conf: 0, expo: 0, publishTime: 0});
```
Comment (anihamde) [inline target_chains/ethereum/sdk/solidity/PriceLibrary.sol:284]: throw error

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+        PythStructs.Price memory self,
+        uint64 deposits,
+        uint64 depositsEndpoint,
+        uint64 rateDiscountInitial,
+        uint64 rateDiscountFinal,
+        int32 discountExponent
+    ) public pure returns (PythStructs.Price memory) {
+        // If the initial discount rate is greater than the final rate, return zero price
+        if (rateDiscountInitial < rateDiscountFinal) {
+            return PythStructs.Price(0, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (nirala-mehul) [reply target_chains/ethereum/sdk/solidity/PriceLibrary.sol]: > Being explicit here with the output would help protect users from incorrectly using an invalid price
Sure added a require check.

File: target_chains/ethereum/sdk/solidity/PriceLibrary.sol (hunk 1 above)
Comment (nirala-mehul) [reply target_chains/ethereum/sdk/solidity/PriceLibrary.sol]: Sure updated it.

Comment (nirala-mehul) [reply target_chains/ethereum/sdk/solidity/PriceLibrary.sol]: > would not do this. `confidence` could still be nonzero

Sure updated it.

**Comment by ali-bahjati** [comment]
We have decided to move forward with #1746

**Pull request closed without merging**